### PR TITLE
Add dynamic search bar

### PR DESCRIPTION
**Pull request by erdemyerebasmaz**
This PR addresses [POS-41](https://breeztech.atlassian.net/browse/POS-41).

A search bar that filters items list as you enter input is added to items view. The items that contains the entered filter are displayed in the items list.
#### TODO:
- ~Enable search by SKU after merging #296~ abecca8
- ~Display no result text for empty results~ 7fa6292  & 9575d52
- ~Add an indicator that the text is filtered~
  - ~Swap search icon with delete icon button that clears filter~89a0d89
- ~Stick search bar to top.~ 3ecd270

### CODE REVIEW

File: lib/routes/charge/items/items_list.dart [hunk 1]
```diff
@@ -21,8 +22,16 @@ class ItemsList extends StatelessWidget {
...
-          return CatalogItem(accountModel, posCatalogBloc, _items[index],
-              _items.length - 1 == index, _addItem);
+          return _filter.isNotEmpty && _containsItem(index) || _filter.isEmpty
+              ? CatalogItem(accountModel, posCatalogBloc, _items[index],
+                  _items.length - 1 == index, _addItem)
+              : Container();
         });
   }
+
+  _containsItem(int index) {
```
Comment (roeierez) [inline lib/routes/charge/items/items_list.dart]: IMO filtering should be done at a lower level (Repository).
We should add action to SetFilter and the items should be fetched from the repository using this filter.

Comment (erdemyerebasmaz) [reply lib/routes/charge/items/items_list.dart]: Filtering is being done at repository layer via FilterSales action.
cc13b598dccd453862b5399c9cb6809a4986517c

File: lib/routes/charge/pos_invoice.dart [hunk 2]
```diff
@@ -56,6 +58,15 @@ class POSInvoiceState extends State<POSInvoice> {
 
   @override
   void didChangeDependencies() {
+    PosCatalogBloc posCatalogBloc =
+        AppBlocsProvider.of<PosCatalogBloc>(context);
+    _itemFilterController.addListener(
+      () => setState(() {
+        FilterSales filterSales = FilterSales(_itemFilterController.text);
+        posCatalogBloc.actionsSink.add(filterSales);
+        filterSales.future.then((_) {});
```
Comment (roeierez) [inline lib/routes/charge/pos_invoice.dart]: We don't need this line.

File: lib/bloc/pos_catalog/repository.dart [hunk 3]
```diff
@@ -13,6 +13,7 @@ abstract class Repository {
   Future<Item> fetchItemByID(int id);
   //This should be converted later to support pagination.
   Future<List<Item>> fetchAllItems();
+  Future<List<Item>> fetchItemsByFilter(String filter);
```
Comment (roeierez) [inline lib/bloc/pos_catalog/repository.dart]: Isn't it better to consolidate these two interface functions to one with optional filter?
`fetchItems({Sting filter}`

File: lib/bloc/pos_catalog/bloc.dart [hunk 4]
```diff
@@ -38,6 +41,14 @@ class PosCatalogBloc with AsyncActionsHandler {
     _itemsStreamController.add(await _repository.fetchAllItems());
   }
 
+  Future _filterSales(FilterSales action) async {
+    if (action.filter.isNotEmpty)
```
Comment (roeierez) [inline lib/bloc/pos_catalog/bloc.dart]: IMO can be more readable if we pass filter parameter to `loadItems`

File: lib/bloc/pos_catalog/actions.dart [hunk 5]
```diff
@@ -43,4 +42,10 @@ class SetCurrentSale extends AsyncAction {
   final Sale currentSale;
 
   SetCurrentSale(this.currentSale);
-}
\ No newline at end of file
+}
+
+class FilterSales extends AsyncAction {
```
Comment (roeierez) [inline lib/bloc/pos_catalog/actions.dart]: FilterItems ?

File: lib/routes/charge/pos_invoice.dart (hunk 2 above)
Comment (erdemyerebasmaz) [reply lib/routes/charge/pos_invoice.dart]: Sorry, that's left from debugging...

File: lib/bloc/pos_catalog/repository.dart (hunk 3 above)
Comment (erdemyerebasmaz) [reply lib/bloc/pos_catalog/repository.dart]: [Done](https://github.com/breez/breezmobile/pull/298/commits/abecca857f6d8dc694c3b74dc617f434a1906501)
addresses renaming comments too.

**Pull request merged**